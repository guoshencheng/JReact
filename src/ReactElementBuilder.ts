import { MaybeArray, ComponentJson } from './JsonReactTypes';

class ReactElementBuilder {
  static build(json: undefined): undefined;
  static build(json: ComponentJson[]) : React.ReactElement<any>[];
  static build(json: ComponentJson): React.ReactElement<any>;
  static build(json: MaybeArray<ComponentJson> | undefined): MaybeArray<React.ReactElement<any>> | undefined {
       
  }
}

export default ReactElementBuilder;