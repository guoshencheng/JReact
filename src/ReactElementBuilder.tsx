import * as React from 'react';
import { Connect } from 'react-redux';
import { MaybeArray, ComponentJson, StringMap, JRComponent, JREvent } from './JsonReactTypes';

class ReactElementBuilder {

  static connect: Connect

  static Components: StringMap<JRComponent> = {}
  static Events: StringMap<JREvent> = {}

  static RegisterComponent(name: string, jrComp: JRComponent) {
    ReactElementBuilder.Components[name] = jrComp;
  }

  static buildMulti(json: ComponentJson[]): (React.ReactElement<any> | string)[] {
    const ele = json.map((item, index) => (
      ReactElementBuilder.buildSingle(item, `${index}`)
    ));
    return ele;
  }

  static buildSingle(json: ComponentJson | string, key?: string): React.ReactElement<any> | string {
    if (typeof json === 'string') {
      return json;
    }
    const { Components } = ReactElementBuilder;
    const { type: t, props, children, data } = json;
    const childNode = ReactElementBuilder.build(children as any);
    // resolve event props
    const component = Components[t];
    let Cls;
    if (!component) {
      // use react origin html tag if component is not found
      Cls = t as keyof React.ReactHTML;
    } else {
      // use registed class or function wraped with connect
      Cls = ReactElementBuilder.connect ? ReactElementBuilder.connect(data, dispatch => ({ dispatch }))(component.Cls) : component.Cls
    }
    return (
      <Cls key={key} {...props}>{ childNode }</Cls>
    )
  }

  static build(json: string): string;
  static build(json?: undefined): undefined;
  static build(json: ComponentJson[]) : React.ReactElement<any>[];
  static build(json: (ComponentJson | string)[]) : (React.ReactElement<any> | string)[];
  static build(json: ComponentJson): React.ReactElement<any>;
  static build(json?: MaybeArray<ComponentJson | string> | undefined): MaybeArray<React.ReactElement<any> | string> | undefined {
    if (!json) return json;
    if (!Array.isArray(json)) {
      return ReactElementBuilder.buildSingle(json);
    }
    return ReactElementBuilder.buildMulti(json);
  }
}

export default ReactElementBuilder;