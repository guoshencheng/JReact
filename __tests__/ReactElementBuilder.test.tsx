import * as React from 'react';
import * as renderer from 'react-test-renderer';
import ReactElementBuilder from '../src/ReactElementBuilder';

describe('check json react render success', () =>  {
  it('render null return undefined', () => {
    const el = ReactElementBuilder.build();
    expect(el).toBeUndefined();
  })
  it('render a array type of json', () => {
    const el = ReactElementBuilder.build([{
      type: 'div',
      props: { style: { left: 20 } }
    }, {
      type: 'div',
      props: { style: { left: 50 } }
    }])
    const tree = renderer.create(
      <React.Fragment>
      {el}
      </React.Fragment>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('render div success', () => {
    const el = ReactElementBuilder.build({
      type: 'div',
      props: { style: { left: 20 } }
    })
    const tree = renderer.create(
      el
    ).toJSON();
    expect(tree).toMatchSnapshot();
  }) 
  it('render div.parent>div.child success', () => {
    const el = ReactElementBuilder.build({
      type: 'div',
      props: { style: { left: 20 }, className: 'parent' },
      children: {
        type: 'div',
        props: {
          className: 'child'
        }
      }
    })
    const tree = renderer.create(el).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('render custom component success', () => {
    const Custom = (props: any) => (
      <div className='custom-component' { ...props }>
        <label>title</label>
        <input type='text'/>
      </div>
    );
    ReactElementBuilder.RegisterComponent('Custom', {
      Cls: Custom
    })
    const el = ReactElementBuilder.build({
      type: 'Custom',
      props: {
        style:{left: 20}
      }
    })
    const tree = renderer.create(el).toJSON();
    expect(tree).toMatchSnapshot();
  })
});

describe('check event handler', () => {
  
})


