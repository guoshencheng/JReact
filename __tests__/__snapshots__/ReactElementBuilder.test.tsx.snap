// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check json react render success render a array type of json 1`] = `
Array [
  <div
    style={
      Object {
        "left": 20,
      }
    }
  />,
  <div
    style={
      Object {
        "left": 50,
      }
    }
  />,
]
`;

exports[`check json react render success render custom component success 1`] = `
<div
  className="custom-component"
  style={
    Object {
      "left": 20,
    }
  }
>
  <label>
    title
  </label>
  <input
    type="text"
  />
</div>
`;

exports[`check json react render success render div success 1`] = `
<div
  style={
    Object {
      "left": 20,
    }
  }
/>
`;

exports[`check json react render success render div.parent>div.child success 1`] = `
<div
  className="parent"
  style={
    Object {
      "left": 20,
    }
  }
>
  <div
    className="child"
  />
</div>
`;
